### PR TITLE
Enable NSXT opaque network support for bridge and container network

### LINKED ISSUE #8317: Enable the opaque network support in VIC CLI
@yuyangbj commented on [Wed Sep 26 2018](https://github.com/vmware/vic-planning/issues/123)

When using vic-machine to create VCH, the VIC could choose opaque network shown in vCenter for bridge and container network.

Add opaque network check with the specified VCH compute cluster.

VM could get up and ip address successfully no matter dhcp/static mode.

Add integration tests.

### PR DESCRIPTION
**Pull request by yuyangbj**
Allow VIC to choose the opaque network for bridge and container
network.

Fixes #8317

### CODE REVIEW

File: lib/install/validate/network.go [hunk 1]
```diff
@@ -412,9 +414,11 @@ func (v *Validator) network(op trace.Operation, input *data.Data, conf *config.V
 			v.NoteIssue(errors.Errorf("the bridge network must not be shared with another network role - %q also mapped as container network %q", input.BridgeNetworkName, name))
 		}
 
-		err = v.checkVDSMembership(op, moref, net)
-		if err != nil && checkMappedVDS {
-			v.NoteIssue(fmt.Errorf("Unable to check hosts in vDS for %q: %s", net, err))
+		if _, dpg := cnet.(object.DistributedVirtualPortgroup); dpg {
```
[the remark this answers is not in the document]
Comment (yuyangbj) [reply lib/install/validate/network.go]: I will try.

**Comment by yuyangbj** [comment]
@zjs thanks for the review, in this PR I only added unit test for networking, but it does not cover opaque network due to govmomi simulator limitation, I will contribute govmomi then. For scenario/nightly test, there is another issue #8331 created, it is planned into next sprint.

File: lib/install/validate/network.go [hunk 2]
```diff
@@ -501,24 +501,26 @@ func (v *Validator) dpgMorefHelper(op trace.Operation, ref string) (string, erro
 	return ref, nil
 }
 
-func (v *Validator) dpgHelper(op trace.Operation, path string) (types.ManagedObjectReference, error) {
+func (v *Validator) dpgAndOpaqueHelper(op trace.Operation, path string) (object.NetworkReference, error) {
```
Comment (hickeng) [inline lib/install/validate/network.go]: Should this be `*object.NetworkReference` instead? That's what we're type checking in the function.
```suggestion
func (v *Validator) dpgAndOpaqueHelper(op trace.Operation, path string) (*object.NetworkReference, error) {
```

File: lib/install/validate/network.go [hunk 3]
```diff
@@ -533,9 +535,13 @@ func (v *Validator) inDVP(op trace.Operation, host types.ManagedObjectReference,
...
 
-// checkVDSMembership verifes all hosts in the vCenter are connected to the vDS
-func (v *Validator) checkVDSMembership(op trace.Operation, network types.ManagedObjectReference, netName string) error {
-	defer trace.End(trace.Begin(network.Value, op))
+// checkVDSMembership verifies all hosts in the vCenter are connected to the vDS
+func (v *Validator) checkVDSMembership(op trace.Operation, network object.NetworkReference, netName string) error {
+	defer trace.End(trace.Begin(network.Reference().Value, op))
+
+	if _, dpg := network.(object.DistributedVirtualPortgroup); !dpg {
+		return nil
```
Comment (hickeng) [inline lib/install/validate/network.go]: Is there no way to check host connectivity for an OpaqueNetwork? At some point we will want to do so in the `pkg/vsphere/compute/placement:CheckHost` logic.

File: lib/install/validate/network.go (hunk 2 above)
Comment (yuyangbj) [reply lib/install/validate/network.go]: Will fix it!

File: lib/install/validate/network.go (hunk 3 above)
Comment (yuyangbj) [reply lib/install/validate/network.go]: I think I can try, I checked there is host property in the OpaqueNetwork object, so we can add the connectivity validation.

File: lib/install/validate/network.go [hunk 4]
```diff
@@ -331,16 +331,18 @@ func (v *Validator) network(op trace.Operation, input *data.Data, conf *config.V
 		op.Debugf("\tNetwork: %s NetworkEndpoint: %v", net, val)
 	}
 
-	err = v.checkVDSMembership(op, endpointMoref, input.BridgeNetworkName)
-	if err != nil && checkBridgeVDS {
-		v.NoteIssue(fmt.Errorf("Unable to check hosts in vDS for %q: %s", input.BridgeNetworkName, err))
+	if endpointNet != nil {
+		err = v.checkNetworkMembership(op, endpointNet, input.BridgeNetworkName)
+		if err != nil && checkBridgeVDS {
+			v.NoteIssue(fmt.Errorf("Unable to check hosts in vDS for %q: %s", input.BridgeNetworkName, err))
```
Comment (wjun) [inline lib/install/validate/network.go]: vDS or opaque network

File: lib/install/validate/network.go [hunk 5]
```diff
@@ -412,9 +414,11 @@ func (v *Validator) network(op trace.Operation, input *data.Data, conf *config.V
 			v.NoteIssue(errors.Errorf("the bridge network must not be shared with another network role - %q also mapped as container network %q", input.BridgeNetworkName, name))
 		}
 
-		err = v.checkVDSMembership(op, moref, net)
-		if err != nil && checkMappedVDS {
-			v.NoteIssue(fmt.Errorf("Unable to check hosts in vDS for %q: %s", net, err))
+		if cnet != nil {
+			err = v.checkNetworkMembership(op, cnet, net)
+			if err != nil && checkMappedNetwork {
+				v.NoteIssue(fmt.Errorf("Unable to check hosts in vDS for %q: %s", net, err))
```
Comment (wjun) [inline lib/install/validate/network.go]: vDS or opaque network

File: lib/install/validate/network.go (hunk 4 above)
Comment (yuyangbj) [reply lib/install/validate/network.go]: Done!

File: lib/install/validate/network.go (hunk 5 above)
Comment (yuyangbj) [reply lib/install/validate/network.go]: Done!

File: lib/install/validate/validator_test.go [hunk 6]
```diff
@@ -103,7 +104,9 @@ func TestValidator(t *testing.T) {
 		conf := testCompute(ctx, validator, input, t)
 		testTargets(ctx, validator, input, conf, t)
 		testStorage(ctx, validator, input, conf, t)
-		testNetwork(ctx, validator, input, conf, t)
+		if i == 1 {
```
[the remark this answers is not in the document]
Comment (yuyangbj) [reply lib/install/validate/validator_test.go]: I remove this limitation!

**Pull request merged**